File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # binary search
        lp, rp = 0, len(nums) - 1
        while lp < rp:
            mid = (lp + rp) // 2
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                lp = mid + 1
            else:
                rp = mid - 1
        if lp == rp and nums[lp] == target:
            return lp
        return -1

File: Data Structures & Algorithms/search-2d-matrix/submission-1.go
func searchMatrix(matrix [][]int, target int) bool {
    lr, rr := 0, len(matrix) - 1
    for lr < rr {
        mid := lr + rr
        if mid % 2 == 0 {
            mid /= 2
        } else {
            mid = (mid + 1) / 2 
        }
        if matrix[mid][0] == target {
            return true
        }
        if matrix[mid][0] > target {
            rr = mid - 1
        } else {
            lr = mid
        }
    }
    lc, rc := 0, len(matrix[0]) - 1
    for lc <= rc {
        mid := (lc + rc) / 2
        if matrix[lr][mid] == target {
            return true
        }
        if matrix[lr][mid] > target {
            rc = mid - 1
        } else {
            lc = mid + 1
        }
    }
    return false
}
